Don't rerun goals if none of their vars have changed

compiler-errors opened the pull request:
r? @ghost

Alternative to rust-lang/rust#141488. I'm pretty sure that we don't need to re-run the goal at all if the inputs don't change... 🤔

bors commented:
:hourglass: Trying commit 2d718c1601bb6f88717322e520072319a4cca07c with merge 4c02404f6a3e21d721602e21942b9ee0e3175b14...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2d718c1601bb6f88717322e520072319a4cca07c","merge_sha":"4c02404f6a3e21d721602e21942b9ee0e3175b14"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15228509825/job/42835380224)
Build commit: 4c02404f6a3e21d721602e21942b9ee0e3175b14 (`4c02404f6a3e21d721602e21942b9ee0e3175b14`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/15228509825/job/42835380224"},"merge_sha":"4c02404f6a3e21d721602e21942b9ee0e3175b14"} -->

rust-timer commented:
Queued 4c02404f6a3e21d721602e21942b9ee0e3175b14 with parent 3de4f1ccf3873782fae2a3883d029ed3d4542581, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3de4f1ccf3873782fae2a3883d029ed3d4542581&end=4c02404f6a3e21d721602e21942b9ee0e3175b14).
There are currently 2 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~3.1 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (4c02404f6a3e21d721602e21942b9ee0e3175b14): [comparison URL](https://perf.rust-lang.org/compare.html?start=3de4f1ccf3873782fae2a3883d029ed3d4542581&end=4c02404f6a3e21d721602e21942b9ee0e3175b14&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=3de4f1ccf3873782fae2a3883d029ed3d4542581&end=4c02404f6a3e21d721602e21942b9ee0e3175b14&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -               | 0     |
| Regressions ❌ <br /> (secondary)  | 0.6%  | [0.3%, 0.9%]    | 6     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.3%, -0.1%]  | 13    |
| Improvements ✅ <br /> (secondary) | -1.8% | [-17.6%, -0.2%] | 23    |
| All ❌✅ (primary)                 | -0.2% | [-0.3%, -0.1%]  | 13    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3de4f1ccf3873782fae2a3883d029ed3d4542581&end=4c02404f6a3e21d721602e21942b9ee0e3175b14&stat=max-rss)
<details>
<summary>Results (primary 0.8%, secondary 5.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.9%  | [1.8%, 2.0%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 6.7%  | [3.2%, 9.6%]   | 7     |
| Improvements ✅ <br /> (primary)   | -1.2% | [-1.2%, -1.2%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-2.8%, -2.8%] | 1     |
| All ❌✅ (primary)                 | 0.8%  | [-1.2%, 2.0%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3de4f1ccf3873782fae2a3883d029ed3d4542581&end=4c02404f6a3e21d721602e21942b9ee0e3175b14&stat=cycles:u)
<details>
<summary>Results (secondary -5.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -               | 0     |
| Regressions ❌ <br /> (secondary)  | 0.9%  | [0.9%, 0.9%]    | 1     |
| Improvements ✅ <br /> (primary)   | -     | -               | 0     |
| Improvements ✅ <br /> (secondary) | -8.6% | [-11.7%, -5.6%] | 2     |
| All ❌✅ (primary)                 | -     | -               | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=3de4f1ccf3873782fae2a3883d029ed3d4542581&end=4c02404f6a3e21d721602e21942b9ee0e3175b14&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 777.23s -> 776.477s (-0.10%)
**Artifact size**: 366.32 MiB -> 366.38 MiB (0.02%)

compiler-errors committed:
Don't rerun goals if none of its vars have changed

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

compiler-errors committed:
Don't retry in pred_known_to_hold_modulo_regions in new solver, since new solver is more complete

Just a totally unrelated nitpick I'm folding into the PR, since it's
code I'd like for us to prune when the new solver lands.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15236859556/job/42853937400)
Build commit: 7e9d4c67ad1d4d4bcd517d72759241344f1854e5 (`7e9d4c67ad1d4d4bcd517d72759241344f1854e5`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/15236859556/job/42853937400"},"merge_sha":"7e9d4c67ad1d4d4bcd517d72759241344f1854e5"} -->

rust-timer commented:
Queued 7e9d4c67ad1d4d4bcd517d72759241344f1854e5 with parent 3d86494a0d0131c32eb15e3a4b685707b9ff000d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3d86494a0d0131c32eb15e3a4b685707b9ff000d&end=7e9d4c67ad1d4d4bcd517d72759241344f1854e5).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (7e9d4c67ad1d4d4bcd517d72759241344f1854e5): [comparison URL](https://perf.rust-lang.org/compare.html?start=3d86494a0d0131c32eb15e3a4b685707b9ff000d&end=7e9d4c67ad1d4d4bcd517d72759241344f1854e5&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=3d86494a0d0131c32eb15e3a4b685707b9ff000d&end=7e9d4c67ad1d4d4bcd517d72759241344f1854e5&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -               | 0     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.2%, 0.4%]    | 5     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.2%, -0.1%]  | 5     |
| Improvements ✅ <br /> (secondary) | -6.1% | [-18.1%, -2.2%] | 6     |
| All ❌✅ (primary)                 | -0.1% | [-0.2%, -0.1%]  | 5     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3d86494a0d0131c32eb15e3a4b685707b9ff000d&end=7e9d4c67ad1d4d4bcd517d72759241344f1854e5&stat=max-rss)
<details>
<summary>Results (primary 0.8%, secondary -3.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.8%  | [1.6%, 2.0%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.3% | [-1.3%, -1.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.3% | [-4.2%, -2.0%] | 14    |
| All ❌✅ (primary)                 | 0.8%  | [-1.3%, 2.0%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3d86494a0d0131c32eb15e3a4b685707b9ff000d&end=7e9d4c67ad1d4d4bcd517d72759241344f1854e5&stat=cycles:u)
<details>
<summary>Results (secondary -0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -               | 0     |
| Regressions ❌ <br /> (secondary)  | 3.5%  | [1.2%, 4.7%]    | 5     |
| Improvements ✅ <br /> (primary)   | -     | -               | 0     |
| Improvements ✅ <br /> (secondary) | -9.2% | [-12.3%, -6.1%] | 2     |
| All ❌✅ (primary)                 | -     | -               | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=3d86494a0d0131c32eb15e3a4b685707b9ff000d&end=7e9d4c67ad1d4d4bcd517d72759241344f1854e5&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 775.592s -> 776.167s (0.07%)
**Artifact size**: 366.33 MiB -> 366.31 MiB (-0.01%)

compiler-errors commented:
Minor improvement in syn. Some of the non new-solver scenarios are back to neutral; not sure what to make of those, but whatever. I think this is fine to be reviewed now.

lcnr reviewed (commented):
:+1: 

naming-wise I am not totally happy yet :thinking: `CanonicalGoalCacheKey` seems somewhat... misleading/unhelpful

- `GoalStalledOn` maybe? also, add a comment to that struct :3

compiler-errors commented on compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs:
```diff
@@ -115,7 +116,7 @@ where
     pub(super) search_graph: &'a mut SearchGraph<D>,
 
-    nested_goals: Vec<(GoalSource, Goal<I, I::Predicate>)>,
+    nested_goals: Vec<(GoalSource, Goal<I, I::Predicate>, Option<CanonicalGoalCacheKey<I>>)>,
```
For a single use? I don't really feel like that's really beneficial. Notably, this isn't the same tuple of `(Goal, Option<CacheKey>)` that's stored in the fulfillment context.

compiler-errors commented on compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs:
```diff
@@ -147,8 +148,9 @@ pub trait SolverDelegateEvalExt: SolverDelegate {
+        cache_key: Option<CanonicalGoalCacheKey<Self::Interner>>,
     ) -> (
-        Result<(HasChanged, Certainty), NoSolution>,
+        Result<(HasChanged, Certainty, CanonicalGoalCacheKey<Self::Interner>), NoSolution>,
```
Unlike the one below, this one actually does clean up a lot 👍

compiler-errors commented on compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs:
```diff
@@ -148,9 +148,9 @@ pub trait SolverDelegateEvalExt: SolverDelegate {
+        stalled_on: Option<GoalStalledOn<Self::Interner>>,
     ) -> (
-        Result<(HasChanged, Certainty, CanonicalGoalCacheKey<Self::Interner>), NoSolution>,
+        Result<GoalEvaluation<Self::Interner>, NoSolution>,
```
I gave the struct that corresponds to `(HasChanged, Certainty, CanonicalGoalCacheKey<I>)` the name `GoalEvaluation`, which coincidentally was also what we call it in `inspect::GoalEvaluation`. I think this parallel is actually good.

compiler-errors commented:
I introduced a new `GoalEvaluation<'tcx>` struct that holds the response + stalled vars from the `evaluate_goal_*` functions, and renamed the "cache key" struct (whose name stopped being descriptive when I changed the approach) to `GoalStalledOn<'tcx>`.

Threw another random commit into the PR (5893b58f0ec808bb96204dc8001e4d1a0783e25c) b/c I didn't want to open a totally separate PR for just a one liner.

compiler-errors commented:
@bors r=lcnr

bors commented:
:pushpin: Commit f0bcb0e929daab62405e6cff19c6c469faaf1791 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr f0bcb0e929daab62405e6cff19c6c469faaf1791 -->
<!-- homu: {"type":"Approved","sha":"f0bcb0e929daab62405e6cff19c6c469faaf1791","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

compiler-errors commented:
@bors r=lcnr

bors commented:
:pushpin: Commit e2215a8ad9099c35df2de789efd9262c3fd59c65 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr e2215a8ad9099c35df2de789efd9262c3fd59c65 -->
<!-- homu: {"type":"Approved","sha":"e2215a8ad9099c35df2de789efd9262c3fd59c65","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/141442#issuecomment-2908997335) for pull requests below priority 10. This pull request will be tested once the tree is reopened.

bors commented:
:hourglass: Testing commit e2215a8ad9099c35df2de789efd9262c3fd59c65 with merge 40d2563ea200f9327a8cb8b99a0fb82f75a7365c...
<!-- homu: {"type":"BuildStarted","head_sha":"e2215a8ad9099c35df2de789efd9262c3fd59c65","merge_sha":"40d2563ea200f9327a8cb8b99a0fb82f75a7365c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15258920420/job/42919513161)
Approved by: lcnr
Pushing 40d2563ea200f9327a8cb8b99a0fb82f75a7365c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/15258920420/job/42919513161"},"merge_sha":"40d2563ea200f9327a8cb8b99a0fb82f75a7365c"} -->

rust-timer commented:
Finished benchmarking commit (40d2563ea200f9327a8cb8b99a0fb82f75a7365c): [comparison URL](https://perf.rust-lang.org/compare.html?start=9c0bcb514f49cd1e6a30affb2fe4cfca060129a2&end=40d2563ea200f9327a8cb8b99a0fb82f75a7365c&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=9c0bcb514f49cd1e6a30affb2fe4cfca060129a2&end=40d2563ea200f9327a8cb8b99a0fb82f75a7365c&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -               | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.2%, 0.2%]    | 1     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.2%, -0.1%]  | 2     |
| Improvements ✅ <br /> (secondary) | -4.1% | [-17.9%, -0.3%] | 9     |
| All ❌✅ (primary)                 | -0.1% | [-0.2%, -0.1%]  | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9c0bcb514f49cd1e6a30affb2fe4cfca060129a2&end=40d2563ea200f9327a8cb8b99a0fb82f75a7365c&stat=max-rss)
<details>
<summary>Results (primary -0.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.0%  | [1.0%, 1.0%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.5% | [-1.6%, -1.4%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.7% | [-1.6%, 1.0%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9c0bcb514f49cd1e6a30affb2fe4cfca060129a2&end=40d2563ea200f9327a8cb8b99a0fb82f75a7365c&stat=cycles:u)
<details>
<summary>Results (secondary -9.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -               | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -               | 0     |
| Improvements ✅ <br /> (primary)   | -     | -               | 0     |
| Improvements ✅ <br /> (secondary) | -9.1% | [-12.3%, -6.0%] | 2     |
| All ❌✅ (primary)                 | -     | -               | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=9c0bcb514f49cd1e6a30affb2fe4cfca060129a2&end=40d2563ea200f9327a8cb8b99a0fb82f75a7365c&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 777.056s -> 778.13s (0.14%)
**Artifact size**: 366.33 MiB -> 366.42 MiB (0.02%)
